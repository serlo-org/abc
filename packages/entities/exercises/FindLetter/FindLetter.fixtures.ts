// TODO: don't depend on React Native app
import Word from '../../../../src/word';
import { Fixtures } from '../helpers';
import { IProps, IState } from './FindLetter';

const props: IProps = {
  word: new Word('ananas'),
  letter: 'A'
};

const fixtures: Fixtures<IProps, IState> = [
  {
    name: 'all correct letters',
    props,
    state: [true, false, true, false, true, false],
    isCorrect: true
  },
  {
    name: 'one wrong letter',
    state: [true, true, true, false, true, false],
    props,
    isCorrect: false
  },
  {
    name: 'some but not all correct letters',
    state: [true, false, false, false, true, false],
    props,
    isCorrect: false
  }
];

export default fixtures;
