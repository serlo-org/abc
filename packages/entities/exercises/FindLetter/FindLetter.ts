import { map } from 'ramda';

import Word from '../../../../src/word';
import AbstractExercise from '../AbstractExercise';

export interface IProps {
  word: Word;
  letter: string;
}

export type IState = boolean[];

class FindLetter extends AbstractExercise<IProps, IState> {
  public getInitialState() {
    return map(
      () => false,
      this.getProps()
        .word.toString()
        .split('')
    ) as IState;
  }

  public isCorrect(state: IState) {
    const { letter, word } = this.getProps();
    const wordString = word.toString().toUpperCase();

    for (let i = 0; i < wordString.length; i++) {
      if (state[i] !== (wordString[i] === letter)) {
        return false;
      }
    }

    return true;
  }
}

export default FindLetter;
