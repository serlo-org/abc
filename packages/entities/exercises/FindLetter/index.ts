import Exercise from './FindLetter';
import fixtures from './FindLetter.fixtures';

export { Exercise, fixtures };
