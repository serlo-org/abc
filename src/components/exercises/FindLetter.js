import { addIndex, map } from 'ramda';
import React, { Component } from 'react';
import { TouchableOpacity, View, Text } from 'react-native';

import speakerImage from '../../assets/images/speaker.png';

import withAudio from '../helpers/withAudio';
import { RoundImageWithButton } from '../Components';
const mapIndexed = addIndex(map);

const styles = {
  letter: {
    padding: 5,
    margin: 2,
    borderRadius: 20,
    elevation: 10,
    shadowColor: 'rgba(0, 0, 0, 0.1)',
    shadowOpacity: 1,
    shadowRadius: 0,
    shadowOffset: {
      height: 4,
      width: 4
    }
  },
  highlighted: {
    backgroundColor: '#73DBFF'
  },
  normal: {
    backgroundColor: '#17BCDE'
  }
};

class FindLetter extends Component {
  constructor(props) {
    super(props);
    this.state = {
      highlighted: map(() => false, this.props.text)
    };
  }

  play = () => {
    const playAll = ([sound, ...rest]) => {
      sound.playAsync();
      sound.setPlaybackFinishedCallback(() => {
        sound.setPositionAsync(0);
        if (rest.length > 0) {
          setTimeout(() => playAll(rest), 1000);
        }
      });
    };
    playAll(this.props.sounds);
  };

  toggleLetter = key => {
    const highlighted = [...this.state.highlighted];
    highlighted[key] = !highlighted[key];
    this.setState({
      highlighted: highlighted
    });
  };

  render() {
    const letters = mapIndexed(
      (char, key) => (
        <TouchableOpacity
          key={key}
          onPress={() => this.toggleLetter(key)}
          style={[
            this.state.highlighted[key] ? styles.highlighted : styles.normal,
            styles.letter
          ]}
        >
          <Text style={{ color: '#fff', fontSize: 40, fontWeight: 'bold' }}>
            {char}
          </Text>
        </TouchableOpacity>
      ),
      this.props.text
    );

    return (
      <View
        style={{
          flex: 1,
          backgroundColor: '#00B4D5',
          alignItems: 'center',
          justifyContent: 'space-around'
        }}
      >
        <RoundImageWithButton
          image={this.props.image}
          imageSize={200}
          icon={speakerImage}
          buttonSize={40}
          onPress={this.play}
        />
        <View style={{ flexDirection: 'row' }}>
          {letters}
        </View>
      </View>
    );
  }
}

export default withAudio(FindLetter);
