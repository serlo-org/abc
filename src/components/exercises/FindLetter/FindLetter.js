import { addIndex, map } from 'ramda';
import React, { Component } from 'react';
import { TouchableOpacity, View, Text } from 'react-native';

import { playAll } from '../../../helpers/audio';
import {
  BLACK_TRANSPARENT,
  PRIMARY_WEAK,
  PRIMARY_STRONG,
  GREEN,
  RED
} from '../../../styles/colors';
import { DEFAULT } from '../../../styles/text';
import WordImageWithSounds from '../../common/WordImageWithSounds';
import { PortraitScreenOrientation } from '../../helpers/screen-orientation';

const mapIndexed = addIndex(map);

const styles = {
  letter: {
    backgroundColor: PRIMARY_STRONG,
    padding: 5,
    margin: 2,
    borderRadius: 20,
    borderWidth: 2,
    elevation: 10,
    shadowColor: BLACK_TRANSPARENT,
    shadowOpacity: 1,
    shadowRadius: 0,
    shadowOffset: {
      height: 4,
      width: 4
    }
  }
};

class FindLetter extends Component {
  play = () => playAll(this.props.sounds);

  toggleLetter = key => () => {
    this.props.setState(state => {
      state[key] = !state[key];

      return state;
    });
  };

  createLetterButton = (letter, i) => {
    const { showFeedback, feedback } = this.props;
    const highlighted = this.props.state[i];
    const wrong = showFeedback && feedback.wrongChoices[i];
    const correct = showFeedback && feedback.correctChoices[i];
    const missingCorrect = showFeedback && feedback.missingCorrectChoices[i];

    return (
      <TouchableOpacity
        key={i}
        onPress={this.toggleLetter(i)}
        style={[
          styles.letter,
          highlighted && {
            backgroundColor: correct ? GREEN : PRIMARY_WEAK
          },
          wrong && {
            backgroundColor: RED
          },
          {
            borderColor: !highlighted && missingCorrect ? GREEN : 'transparent'
          }
        ]}
      >
        <Text style={DEFAULT}>{letter}</Text>
      </TouchableOpacity>
    );
  };

  render() {
    const letters = mapIndexed(
      this.createLetterButton,
      this.props.word.toString()
    );

    return (
      <PortraitScreenOrientation>
        <View
          style={{
            flex: 1,
            alignItems: 'center',
            justifyContent: 'space-around'
          }}
        >
          <WordImageWithSounds playInitially word={this.props.word} longSound />
          <View style={{ flexDirection: 'row' }}>{letters}</View>
        </View>
      </PortraitScreenOrientation>
    );
  }
}

export default FindLetter;
