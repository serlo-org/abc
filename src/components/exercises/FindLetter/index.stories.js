import { storiesOf } from '@storybook/react-native';

import * as FindLetter from '.';
import { createStoriesFromFixtures } from '../helpers';

const story = storiesOf('exercises/FindLetter', module);

createStoriesFromFixtures(story, FindLetter);
