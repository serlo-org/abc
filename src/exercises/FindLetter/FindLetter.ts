import { map } from 'ramda';

import { getWord } from '../../helpers/words.js';
import AbstractExercise from '../AbstractExercise';

export interface IProps {
  word: any;
  letter: string;
}

export type IState = boolean[];

class FindLetter extends AbstractExercise<IProps, IState> {
  public getInitialState() {
    return map(() => false, getWord(this.getProps().word)) as IState;
  }

  public isCorrect(state: IState) {
    const { letter, word } = this.getProps();
    const wordString = getWord(word).toUpperCase();

    for (let i = 0; i < wordString.length; i++) {
      if (state[i] !== (wordString[i] === letter)) {
        return false;
      }
    }

    return true;
  }
}

export default FindLetter;
