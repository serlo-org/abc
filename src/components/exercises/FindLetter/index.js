import { FindLetter } from '../../../../packages/entities/exercises';
import Component from './FindLetter';

export const Exercise = FindLetter.Exercise;
export const fixtures = FindLetter.fixtures;
export { Component };
