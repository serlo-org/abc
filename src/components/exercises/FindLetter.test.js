import React from 'react';

import FindLetter from './FindLetter';

import renderer from 'react-test-renderer';

it('renders without crashing', () => {
  const tree = renderer.create(
    <FindLetter
      image={require('../../assets/images/ananas.jpg')}
      sounds={[
        require('../../assets/sounds/ananas_short.mp3'),
        require('../../assets/sounds/ananas_long.mp3')
      ]}
      text="Ananas"
    />
  );
  expect(tree).toMatchSnapshot();
});
