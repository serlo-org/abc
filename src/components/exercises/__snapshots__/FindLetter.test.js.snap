// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#00b4d5",
      "flex": 1,
      "justifyContent": "space-around",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flexDirection": "row",
      }
    }
  >
    <View
      collapsable={undefined}
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.1)",
          "borderRadius": 9999999,
          "margin": 20,
        }
      }
    >
      <Image
        collapsable={undefined}
        resizeMode="cover"
        source={1337}
        style={
          Object {
            "borderRadius": 100,
            "height": 200,
            "margin": 20,
            "width": 200,
          }
        }
      />
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginLeft": -70,
          "marginRight": 20,
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Object {
            "backgroundColor": "#73dbff",
            "borderRadius": 9999,
            "elevation": 10,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
          }
        }
      >
        <Image
          source={1337}
          style={
            Object {
              "height": 40,
              "resizeMode": "contain",
              "width": 40,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#17bcde",
          "borderRadius": 20,
          "elevation": 10,
          "margin": 2,
          "opacity": 1,
          "padding": 5,
          "shadowColor": "rgba(0, 0, 0, 0.1)",
          "shadowOffset": Object {
            "height": 4,
            "width": 4,
          },
          "shadowOpacity": 1,
          "shadowRadius": 0,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "norddruck",
            "fontSize": 35,
            "textAlign": "center",
          }
        }
      >
        A
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#17bcde",
          "borderRadius": 20,
          "elevation": 10,
          "margin": 2,
          "opacity": 1,
          "padding": 5,
          "shadowColor": "rgba(0, 0, 0, 0.1)",
          "shadowOffset": Object {
            "height": 4,
            "width": 4,
          },
          "shadowOpacity": 1,
          "shadowRadius": 0,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "norddruck",
            "fontSize": 35,
            "textAlign": "center",
          }
        }
      >
        n
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#17bcde",
          "borderRadius": 20,
          "elevation": 10,
          "margin": 2,
          "opacity": 1,
          "padding": 5,
          "shadowColor": "rgba(0, 0, 0, 0.1)",
          "shadowOffset": Object {
            "height": 4,
            "width": 4,
          },
          "shadowOpacity": 1,
          "shadowRadius": 0,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "norddruck",
            "fontSize": 35,
            "textAlign": "center",
          }
        }
      >
        a
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#17bcde",
          "borderRadius": 20,
          "elevation": 10,
          "margin": 2,
          "opacity": 1,
          "padding": 5,
          "shadowColor": "rgba(0, 0, 0, 0.1)",
          "shadowOffset": Object {
            "height": 4,
            "width": 4,
          },
          "shadowOpacity": 1,
          "shadowRadius": 0,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "norddruck",
            "fontSize": 35,
            "textAlign": "center",
          }
        }
      >
        n
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#17bcde",
          "borderRadius": 20,
          "elevation": 10,
          "margin": 2,
          "opacity": 1,
          "padding": 5,
          "shadowColor": "rgba(0, 0, 0, 0.1)",
          "shadowOffset": Object {
            "height": 4,
            "width": 4,
          },
          "shadowOpacity": 1,
          "shadowRadius": 0,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "norddruck",
            "fontSize": 35,
            "textAlign": "center",
          }
        }
      >
        a
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#17bcde",
          "borderRadius": 20,
          "elevation": 10,
          "margin": 2,
          "opacity": 1,
          "padding": 5,
          "shadowColor": "rgba(0, 0, 0, 0.1)",
          "shadowOffset": Object {
            "height": 4,
            "width": 4,
          },
          "shadowOpacity": 1,
          "shadowRadius": 0,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "norddruck",
            "fontSize": 35,
            "textAlign": "center",
          }
        }
      >
        s
      </Text>
    </View>
  </View>
</View>
`;
